### DISCUSSION TITLE
Topiary for Markdown code blocks

### DISCUSSION THREAD

**Discussion by aspiwack** [topiary/topiary#400] [Ideas]
I don't remember, maybe we've had this conversation already. Or maybe not. But at any rate, now it's public, so all the better. Not making an issue, because it's just a silly half-baked idea.

Is it plausible to use Topiary to format the code blocks in a Markdown file, using a grammar that would recognise which grammar to delegate each code block?

There are two existing Markdown grammars referenced on the Tree-Sitter website https://github.com/ikatyang/tree-sitter-markdown and https://github.com/MDeiml/tree-sitter-markdown . I don't know if either of them deals with code blocks.

Even if they did, there would be an interesting challenge in using the appropriate Topiary query file to deal with the code block: we're not going to copy them in the query file for Markdown.

Editors that support syntax colouring inside Markdown tend to do so by switching mode (Emacs terminology) between code block fence. In our world, it would mean that, instead of using Topiary to format a Markdown file, we'd use a different executable (possibly Topiary, but maybe another, like Gofmt) for each code block. How do the approaches compare?

**Comment by torhovland** [comment]
That's an interesting idea. There's no reason why Topiary couldn't take the language from a ```` ```language```` code block and format the block if the language is supported. Similarly, it should be possible to format the rest of the Markdown with a Markdown grammar and treat any code blocks as leaf nodes.

**Comment by Lord-Valen** [comment]
This would be interesting for org-mode code blocks as well.

**Reply by toastal** [reply]
And AsciiDoc, Djot, reStructuredText

**Comment by ErinvanderVeen** [comment]
For future reference, these things are often called "injections" in the tree-sitter community: https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection

**Reply by reteps** [reply]
New link: https://tree-sitter.github.io/tree-sitter/3-syntax-highlighting.html#language-injection
